sparse/file.go

```diff
@@ -192,46 +194,186 @@ func HashData(data []byte) ([]byte, error) {
 	return sum[:], nil
 }
 
-func GetFiemapExtents(file FileIoProcessor) ([]Extent, error) {
+func GetFileLayout(ctx context.Context, file FileIoProcessor) (<-chan FileInterval, <-chan error, error) {
 	fileInfo, err := file.Stat()
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-	return GetFiemapRegionExts(file, Interval{0, fileInfo.Size()})
+
+	const MaxExtentsBuffer = 1024
+	const MaxInflightIntervals = MaxExtentsBuffer * 10
+	out := make(chan FileInterval, MaxInflightIntervals)
+	errc := make(chan error, 1)
+	go func() {
+		defer close(out)
+		defer close(errc)
+
+		startTime := time.Now()
+		totalExtents := 0
+		defer func() {
+			log.Debugf("retrieved extents for file %v, fileSize: %v elapsed: %.2fs, extents: %v",
+				fileInfo.Name(), fileInfo.Size(),
+				time.Now().Sub(startTime).Seconds(), totalExtents)
+		}()
+
+		var lastIntervalEnd int64
+		isFinalInterval := fileInfo.Size() == 0
+		for !isFinalInterval {
+			exts, err := GetFiemapRegionExts(file, Interval{lastIntervalEnd, fileInfo.Size()}, MaxExtentsBuffer)
+			if err != nil {
+				errc <- err
+				return
+			}
+
+			intervals := make([]FileInterval, 0, len(exts))
+			isFinalInterval = len(exts) == 0
+			totalExtents += len(exts)
+
+			// special case the whole file is a hole
+			if totalExtents == 0 && fileInfo.Size() != 0 {
+				hole := FileInterval{SparseHole, Interval{0, fileInfo.Size()}}
+				intervals = append(intervals, hole)
+			}
+
+			// map to intervals
+			for _, e := range exts {
+				data := FileInterval{SparseData, Interval{int64(e.Logical), int64(e.Logical + e.Length)}}
+
+				if lastIntervalEnd < data.Begin {
+					hole := FileInterval{SparseHole, Interval{lastIntervalEnd, data.Begin}}
+					intervals = append(intervals, hole)
+				}
+
+				lastIntervalEnd = data.End
+				intervals = append(intervals, data)
+				isFinalInterval = e.Flags&FIEMAP_EXTENT_LAST != 0
+
+				if isFinalInterval {
+					// add a hole between last data segment and end of file
+					if lastIntervalEnd < fileInfo.Size() {
+						hole := FileInterval{SparseHole, Interval{lastIntervalEnd, fileInfo.Size()}}
+						intervals = append(intervals, hole)
+					}
+				}
+			}
+
+			// transmit intervals
+			for _, interval := range intervals {
```

Comment:
I am thinking if this will degrade the performance a lillte bit since this is like a barrier in the med of a pipeline. 
For example, the below signal sending is suddenly blocked for a while, everything will get stuck. But actually, this goroutine can continue filling up the interval channel as long as the channel buffer is not full.

Reply:
Another fair point, good that you asked. Let me elaborate on why I picked this design.

The base assumption / case here is that the rate of **inflow > outflow**, 
- This holds since the time requirement of processing a interval is much higher than the retrieval of the extents.
- Another thing to note is that the processing of intervals is io bound and not cpu bound.
- Which makes it a good candidate for concurrency to begin with.

If the rate of inflow < outflow then a trickle down pipeline (push individual intervals) might be better, but in that case one should also look into reducing the outflow (worker - 1) to further reduce the wait time.

------------------------------------------------------------------------------------------------

So why this design? (of course the analogy has some leaky abstractions :)
Here is an analogy, we have a 10L bucket, with 4 holes at the bottom. 
If we let water slowly drip into it (individual segments) then there will be a lot more contention at the beginning but the startup delay is reduced since each drip can go out of a single hole, while the 3 other holes wait for the next drip. 
If instead we fill up a 1L bottle first then dump all that water in at once, all the holes will have a continuous outflow (even if there is variance at the rate of outflow). This comes with a small startup delay (first 1024 extents) but in our case doesn't matter.

------------------------------------------------------------------------------------------------

Enough with the analogies, let's talk about the code:
- splits the concerns of map(extent -> interval) and reduce (drain interval)
- makes it easier to read the code, since special hole case / hole / data intervals all end up using the same code path.
- single dedicated place where we write into the channel, this means there is also only a single place where the routine can block. This also means that when the channel gets unblocked, there is only a single place for the scheduler to return and the next item will be immediately pushed back into the channel.
- this allows the cpu to cache the interval buffer, since there are no more writes to it.
- this separation makes it really easy to split/merge data intervals if we want to add later.
- we want to have a buffered channel, I used (10 * MAX_EXTENTS) so we have bounded memory consumption

---------------------------------------------------------------------------

These are some of the reasons for the above design. Using a trickle down style pipeline would be just as valid, but would make the code less easy to read, since there would be many places where you could end up blocking (buffer full) this makes it harder to benchmark/analyse a runtime profile.

Let me know what you think :)

Reply:
Got it. We can handle the production and the consumption separately if the delay of the production can be ignored.
And the benefits look good. But I don't understand these 2 parts, can you explain more about them?
1. `there is only a single place for the scheduler to return and the next item will be immediately pushed back into the channel`
2. `this allows the cpu to cache the interval buffer, since there are no more writes to it.`

Reply:
Below are some really good information on how the go scheduler works and should be read in order :)

Simple overview:
https://riteeksrivastava.medium.com/a-complete-journey-with-goroutines-8472630c7f5c#:~:text=The%20switch%20between%20Goroutines%20happen,if%20those%20operations%20would%20block.

Nice gotcha pre GO 1.14
http://www.sarathlakshman.com/2016/06/15/pitfall-of-golang-scheduler

Detailed Overview:
https://medium.com/@ankur_anand/illustrated-tales-of-go-runtime-scheduler-74809ef6d19b

New non cooperative preemptive scheduler GO 1.14:
https://github.com/golang/proposal/blob/master/design/24543-non-cooperative-preemption.md

-----------------------------------------------

This is more of witch craft, so you probably should it ignore it :)
There is also a bunch of leaky abstractions involved here :)

As long as the GO routine P, which would be doing the For Loop: Interval -> sent to channel
Can remain bound to Thread M and continue running for multiple executions instead of being preempted.
Since their are no writes to the underlying buffer, the CPU core that is executing thread M, should be able to load up the underlying array of the intervals slice into the cache instead of having to fetch each element from memory.